src/category_theory/sites/cocontinuous.lean

```diff
@@ -0,0 +1,214 @@
+/-
+Copyright (c) 2021 Andrew Yang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Andrew Yang
+-/
+import category_theory.sites.sheaf
+import category_theory.limits.kan_extension
+
+/-!
+# Cocontinuous functors between sites.
+
+We define cocontinuous functors between sites as functors that pulls covering sieves back to
+covering sieves. This seems stronger that the definition found in stacks project,
+but they are actually equivalent via `category_theory.grothendieck_topology.superset_covering`
+(The precise statement not formalized, but follows from it quite trivially).
+
+This concept is also called the "covering lifting property" as defined in [MM92] Chapter VII,
+Section 10, and should not be confused with the general definition of cocontinuous functors
+between categories as functors preserving small colimits.
+
+## Main Results
+- `category_theory.sites.Ran_is_sheaf_of_cocontinuous`: If `u : C ⥤ D` is cocontinuous, then
+`Ran u.op` (`ₚu`) as a functor `(Cᵒᵖ ⥤ A) ⥤ (Dᵒᵖ ⥤ A)` of presheaves maps sheaves to sheaves.
+
+## References
+
+* [S. MacLane, I. Moerdijk, *Sheaves in Geometry and Logic*][MM92]
+* https://stacks.math.columbia.edu/tag/00XI
+
+-/
+
+universes u
+noncomputable theory
+
+open category_theory
+open opposite
+open category_theory.presieve.family_of_elements
+open category_theory.presieve
+open category_theory.limits
+
+namespace category_theory
+namespace sites
+variables {C D A : Type u} [category.{u} C] [category.{u} D] [category.{u} A] [has_limits A]
+variables {J : grothendieck_topology C} {K : grothendieck_topology D}
+
+/--
+A functor `u : (C, J) ⥤ (D, K)` between sites is called cocontinuous if for all covering sieves
+`R` in `D`, `R.pullback u` is a covering sieve in `C`.
+-/
+structure cocontinuous (J : grothendieck_topology C) (K : grothendieck_topology D) (u : C ⥤ D) :=
+(cover_lift : ∀ {U : C} {S : sieve (u.obj U)} (hS : S ∈ K (u.obj U)), S.functor_pullback u ∈ J U)
+
+/-- A trivial example to make the inhabited linter happy. -/
+instance id_cocontinuous : inhabited (cocontinuous J J (𝟭 _)) :=
```

Comment:
```suggestion
lemma id_cocontinuous : inhabited (cocontinuous J J (𝟭 _)) :=
```

Reply:
Since `cocontinuous` is a structure (instead of `class`) you cannot use instances, but should use `lemma/def` instead.
And I think that using a structure instead of class is the right choice.

Reply:
Why is it? I was just considering if I should change into a class since I would like to be able to write `ₛ[u]` or something similar for `Ran u.op` restricted onto the category of sheaves.

Reply:
The notation is unrelated to the `structure`-vs-`class` design choice, right? I don't have a strong opinion on what you end up using.

For continuous functions we don't use the typeclass system for continuity. One reason is that it has trouble with composition of functions. But composition of functors will probably not cause those issues, because a functor is not a lambda expression.

Additive functors are implemented using a typeclass. So it wouldn't be strange to mimic that choice here.

Reply:
I imagine that for the notation to work, Lean would need to infer the `cocontinuous` stuff by itself, and it seems to be easier to do if `cocontinuous` were a typeclass.
I'm reconsidering the choices on the notations though.
But nonetheless I think it would be better as a `class`.

Reply:
In order to make it a class, I think we might need to bundle the underlying category and the topology into a site and let the functor be between the sites. Or else Lean cannot infer the Grothendieck topology on its own. 

I'm not sure if this is the right thing to do or not.

Reply:
Good point. In that case, I suggest we leave it a structure.